Modules/_datetimemodule.c

```diff
@@ -4839,6 +4852,41 @@ datetime_combine(PyObject *cls, PyObject *args, PyObject *kw)
     return result;
 }
 
+
+static PyObject *
+_sanitize_isoformat_str(PyObject *dtstr, unsigned char *needs_decref) {
+    // `fromisoformat` allows surrogate characters in exactly one position,
+    // the separator; to allow datetime_fromisoformat to make the simplifying
+    // assumption that all valid strings can be encoded in UTF-8, this function
+    // replaces any surrogate character separators with `T`.
+    Py_ssize_t len = PyUnicode_GET_LENGTH(dtstr);
+    *needs_decref = 0;
+    if (len <= 10 || !Py_UNICODE_IS_SURROGATE(PyUnicode_READ_CHAR(dtstr, 10))) {
+        return dtstr;
+    }
+
+    PyObject *left = PyUnicode_Substring(dtstr, 0, 10);
+    if (left == NULL) {
+        return NULL;
+    }
+
+    PyObject *right = PyUnicode_Substring(dtstr, 11, len);
+    if (right == NULL) {
+        Py_DECREF(left);
+        return NULL;
+    }
+
+    PyObject *str_out = PyUnicode_FromFormat("%UT%U", left, right);
```

Comment:
I haven't checked it, but creating a copy of `dtstr` (with `PyUnicode_New`/ `PyUnicode_CopyCharacters` or similar) and using `PyUnicode_WriteChar` to replace the separator might be both faster and simpler than splitting the string and then joining it again.

Reply:
Not sure how to go from "T" to a PyUCS4.

Reply:
OK, refactored. You are right, it is both simpler and faster.

Reply:
Great, thanks!